### PR TITLE
Check return type of generateLabelRecord method

### LINKED ISSUE #4125: generateRecordLabel returns an array instead of string when deleting a DC_Table
### Affected version(s)

4.13 RC3

### Description

Hello,

we've got following issue with one of our plugins when deleting an entry in a DC_Table:

```
Contao\CoreBundle\EventListener\DataContainer\RecordPreviewListener::compilePreview(): Return value must be of type string, array returned

in **\contao413\vendor\contao\core-bundle\src\EventListener\DataContainer\RecordPreviewListener.php (line 89)

```

Contao Version: 4.13 RC3
PHP Version: 8.0.3

Stacktrace:
```
TypeError:
Contao\CoreBundle\EventListener\DataContainer\RecordPreviewListener::compilePreview(): Return value must be of type string, array returned

  at **\contao413\vendor\contao\core-bundle\src\EventListener\DataContainer\RecordPreviewListener.php:89
  at Contao\CoreBundle\EventListener\DataContainer\RecordPreviewListener->compilePreview(object(DC_Table), array('id' => '26', 'tstamp' => '1644791731', 'title' => 'asd', 'desc' => null, 'identifier' => 'asddd', 'groupAlias' => '', 'sorting' => '128'))
     (**\contao413\vendor\contao\core-bundle\src\EventListener\DataContainer\RecordPreviewListener.php:63)
  at Contao\CoreBundle\EventListener\DataContainer\RecordPreviewListener->storePrecompiledRecordPreview(object(DC_Table), '31')
     (**\contao413\vendor\contao\core-bundle\src\Resources\contao\drivers\DC_Table.php:1506)
  at Contao\DC_Table->delete()
     (**\contao413\vendor\contao\core-bundle\src\Resources\contao\classes\Backend.php:655)
  at Contao\Backend->getBackendModule('style_manager', null)
     (**\contao413\vendor\contao\core-bundle\src\Resources\contao\controllers\BackendMain.php:170)
  at Contao\BackendMain->run()
     (**\contao413\vendor\contao\core-bundle\src\Controller\BackendController.php:49)
  at Contao\CoreBundle\Controller\BackendController->mainAction()
     (**\contao413\vendor\symfony\http-kernel\HttpKernel.php:152)
  at Symfony\Component\HttpKernel\HttpKernel->handleRaw(object(Request), 1)
     (**\contao413\vendor\symfony\http-kernel\HttpKernel.php:74)
  at Symfony\Component\HttpKernel\HttpKernel->handle(object(Request), 1, true)
     (**\contao413\vendor\symfony\http-kernel\Kernel.php:202)
  at Symfony\Component\HttpKernel\Kernel->handle(object(Request))
     (**\contao413\public\index.php:44)   

in **\contao413\vendor\contao\core-bundle\src\EventListener\DataContainer\RecordPreviewListener.php (line 89)  
```

The affected DCA:

[https://github.com/oveleon/contao-component-style-manager/blob/2aaec7be9aaafce5b9e[…]fe83505c79213f2c4/src/Resources/contao/dca/tl_style_manager.php](https://github.com/oveleon/contao-component-style-manager/blob/2aaec7be9aaafce5b9e13cdfe83505c79213f2c4/src/Resources/contao/dca/tl_style_manager.php#L8)

(Parent archive: [https://github.com/oveleon/contao-component-style-manager/blob/2aaec7be9aaafce5b9e[…]79213f2c4/src/Resources/contao/dca/tl_style_manager_archive.php](https://github.com/oveleon/contao-component-style-manager/blob/2aaec7be9aaafce5b9e13cdfe83505c79213f2c4/src/Resources/contao/dca/tl_style_manager_archive.php#L12))

### PR DESCRIPTION
**Pull request by bezin**
Fixes #4125

### CODE REVIEW

**Review by fritzmg — CHANGES_REQUESTED** [review]
Before we merge this, I'll check what exactly happens in `generateRecordLabel` with the given DCA.

**Comment by bezin** [comment]
The registered [label_callback returns an array](https://github.com/oveleon/contao-component-style-manager/blob/2aaec7be9aaafce5b9e13cdfe83505c79213f2c4/src/Resources/contao/dca/tl_style_manager_archive.php#L216-L224) and `showColumns` is not set. 

A l[abel_callback that returns an array is not handled](https://github.com/contao/contao/blob/6d8af7b525f85e790b9edb2b6627354f70019040/core-bundle/src/Resources/contao/classes/DataContainer.php#L1773-L1776) in `generateRecordLabel`. It only depends on `showColumns` being true or not (or parent or tree mode).

File: core-bundle/src/EventListener/DataContainer/RecordPreviewListener.php [hunk 1]
```diff
@@ -82,10 +82,7 @@ private function compilePreview(DataContainer $dc, array $row): string
             }
         }
 
-        if ($GLOBALS['TL_DCA'][$dc->table]['list']['label']['showColumns'] ?? false) {
-            return serialize($dc->generateRecordLabel($row, $dc->table));
-        }
-
-        return $dc->generateRecordLabel($row, $dc->table);
+        $preview = $dc->generateRecordLabel($row, $dc->table);
```
Comment (fritzmg) [inline core-bundle/src/EventListener/DataContainer/RecordPreviewListener.php:85]:
```suggestion
        $preview = $dc->generateRecordLabel($row, $dc->table);

```

**Review by fritzmg — DISMISSED** [review]
`DC_Table::listView` automatically handles labels that are arrays if `showColumns` is not set:

https://github.com/contao/contao/blob/90ea2319099c44862d1607c22cd79a7cc762ffab/core-bundle/src/Resources/contao/drivers/DC_Table.php#L4985-L4989

Therefore the listener must not check for `showColumns` itself 👍

File: core-bundle/src/EventListener/DataContainer/RecordPreviewListener.php [hunk 2]
```diff
@@ -82,10 +82,8 @@ private function compilePreview(DataContainer $dc, array $row): string
             }
         }
 
-        if ($GLOBALS['TL_DCA'][$dc->table]['list']['label']['showColumns'] ?? false) {
-            return serialize($dc->generateRecordLabel($row, $dc->table));
-        }
+        $preview = $dc->generateRecordLabel($row, $dc->table);
 
-        return $dc->generateRecordLabel($row, $dc->table);
+        return (is_array($preview)) ? serialize($preview) : $preview;
```
Comment (fritzmg) [inline core-bundle/src/EventListener/DataContainer/RecordPreviewListener.php]:
```suggestion
        return (\is_array($preview)) ? serialize($preview) : $preview;
```

**Comment by fritzmg** [comment]
You should run

```
tools/ecs/vendor/bin/ecs check *-bundle/src/Resources/contao --config tools/ecs/config/legacy.php --fix --ansi
```

locally first, before committing ;)

**Comment by bezin** [comment]
I usually do ;-) All set now.

**Review by fritzmg — DISMISSED** [review]

**Review by aschempp — CHANGES_REQUESTED** [review]
This can't be right. Label callbacks and others are expected to return an array if the data is displayed in a column view. How do we support that now?

**Comment by bezin** [comment]
> This can't be right. Label callbacks and others are expected to return an array if the data is displayed in a column view. How do we support that now?

We still do? I simply do not check for `showColumns = true` specifically as one may return an array from the label callback without setting `showColumns` to `true`. I use whatever the  generateRecordLabel` method returns now.

**Comment by fritzmg** [comment]
@aschempp see https://github.com/contao/contao/pull/4126#pullrequestreview-881286439

**Review by aschempp — COMMENTED** [review]
Ah sorry I somehow thought `generateRecordLabel` would have a string return type 🤦 

However, shouldn't we implode the value like DC_Table does instead of serializing? Or how do we unserialize and correctly handle array values when displaying our undo view?

**Comment by bezin** [comment]
> However, shouldn't we implode the value like DC_Table does instead of serializing? Or how do we unserialize and correctly handle array values when displaying our undo view?

You have point here: Before this PR, the presence of a label as array values equaled `showColumns = true`, so I [rendered that in a table, as it would be in the listing](https://github.com/contao/contao/blob/6d8af7b525f85e790b9edb2b6627354f70019040/core-bundle/src/Resources/views/Backend/be_undo_label.html.twig#L23-L31). 

However, now this does not apply anymore – hence I would suggest to do check for `showColumns = true` after all and 
1. store a serialized array and display it as tabular data, when `showColumns = true`;
2. implode array data and store as string if `generateRecordLabel` returns an array and `showColumns = false`; or
3. store label as string and render as string if `generateRecordLabel` returns a string.

**Review by fritzmg — DISMISSED** [review]

**Comment by aschempp** [comment]
I would think we should always store the array and render depending on `showColumns` (implode or render columns)?

**Comment by bezin** [comment]
> I would think we should always store the array and render depending on `showColumns` (implode or render columns)?

I implemented it that way in [947cbff](https://github.com/contao/contao/pull/4126/commits/947cbffdc7381523a59741e688d7862f23106d7f) @aschempp

**Review by leofeyer — APPROVED** [review]

**Comment by leofeyer** [comment]
Thank you @bezin.

**Pull request merged**